YesterdayApi/Core/Email/Builder/Builders/RegistrationConfirmationEmailBuilder.cs

```diff
@@ -0,0 +1,56 @@
+using Microsoft.AspNetCore.Mvc.RazorPages;
+using MimeKit;
+using RazorHtmlEmails.RazorClassLib.Views;
+using YesterdayApi.Core.Email.Builder.Renderer;
+
+namespace YesterdayApi.Core.Email.Builder.Builders
+{
+    public class RegistrationConfirmationEmailBuilder : EmailBuilderBase
+    {
+        private string _subject;
```

Comment:
This builders are weird. Builders are useful when you create things many time and you would like to avoid long constructors. But here, registration email builder only gonna be use in ONE PLACE which is CreateUser method. For me we need only one builder which is EmailBuilder. Then you use it like this:

```
builder.
   .SetFrom("...") <-- or DopeConstants.Email
   .SetTo(user.Email)
   .SetSubject("...") <-- or RegisterEmail.Subject
   .SetBody(templatePath, templateModel)
   .Build(); <-- return MimeMessage
```

SetBody should use template engine to pass model to html.

Or if you would like to avoid magic strings ("...") you can always create constant class like RegisterEmailConstants (and use it like RegisterEmailConstants.TemplatePath)

Reply:
Builder allows you to produce different types and representations of an object using the same construction code. Our product here is a email in this case.

The director class defines the order in which to execute the building steps, while the builder provides the implementation for those steps.

There is a difference between a registration email / reset password email / newsletter email. I.e. registration / reset pass email you are going to generate some kind of token / link which you wouldn't need in a newsletter email, thats why you need a variety of builders with have different implementations.

The client should decide what kind of email he wants to create thats why we pass the **Type** to the director and let the factory method **PrepareBuilder** decide what kind of builder to create. Other approach to do that would be to create a builder by the client and pass it to the director with would use it. But we ( / I ) would not want to bother the client with it.

Reply:
> The director class defines the order in which to execute the building steps, while the builder provides the implementation for those steps.

Isn't that template method? Just curious xd

Reply:
> > The director class defines the order in which to execute the building steps, while the builder provides the implementation for those steps.
> 
> Isn't that template method? Just curious xd

Template Method is a behavioral design pattern that defines the skeleton of an algorithm in the superclass but lets subclasses override specific steps of the algorithm without changing its structure.